### PR TITLE
[SIL]/[SILOpt] Thunks and specializations are never public.

### PR DESCRIPTION
**Pull request by huonw**
The current behaviour of (some) open classes means private members are
forced to be public, but this only applies to the real members, not
thunks or specializations associated with them.

Fixes rdar://problem/40738913.

### CODE REVIEW

**Comment by huonw** [comment]
@swift-ci please test

File: include/swift/SIL/SILFunction.h [hunk 1]
```diff
@@ -430,6 +430,11 @@ class SILFunction
   /// method itself, the function can be referenced from vtables of derived
   /// classes in other compilation units.
   SILLinkage getEffectiveSymbolLinkage() const {
+    // These are implementations details that don't get special treatment: their
+    // stored linkage is the correct one.
+    if (isThunk() || isSpecialization())
```
Comment (slavapestov) [inline include/swift/SIL/SILFunction.h]: I think it would make more sense to turn this into an assertion and fix those places that create thunks no?

**Review by slavapestov — CHANGES_REQUESTED** [review]

**Comment by huonw** [comment]
Linux failure (for future reference):

```
20:53:10 FAILED: ../buildbot_linux/foundation-linux-x86_64/Foundation/libFoundation.so 
...
../buildbot_linux/foundation-linux-x86_64/Foundation/Foundation/NSCFDictionary.swift.o:NSCFDictionary.swift.o:$S10Foundation15_NSCFDictionaryCMf: error: undefined reference to '$S10Foundation12NSDictionaryC14getDescription33_8C53A72A73712D943F7602D2573DFCFBLL2ofSSSgyp_tF'
```

File: include/swift/SIL/SILFunction.h (hunk 1 above)
Comment (huonw) [reply include/swift/SIL/SILFunction.h]: Hm, I'm not sure: this is called in places that have just an arbitrary `SILFunction`:

https://github.com/apple/swift/blob/e8ee93b131d2ad360a8ff02034eb6c29718e411c/lib/IRGen/GenDecl.cpp#L1618-L1619

We could push this condition down to them, but I'm not sure it makes sense. To me, it seems reasonable to say that `getEffectiveSymbolLinkage` is the way to get the true IR/ELF/mach-O linkage, and thus having this split here is the right place.

Comment (slavapestov) [reply include/swift/SIL/SILFunction.h]: I'm suggesting that `getClassSubclassScope()` should not be set on thunks. Whoever is creating the thunk's SILFunction is also setting `getClassSubclassScope()`, which seems clearly wrong.

Comment (huonw) [reply include/swift/SIL/SILFunction.h]: Ah, I had misunderstood. That sounds good.

**Comment by jrose-apple** [comment]
Mm. What about override thunks, for concrete subclasses of generic classes?

```
open class Base<T> {
  open func consume(_: T) {}
}

open class Sub: Base<String> {
  open override func consume(_: String) { print("override") }
}
```

**Comment by slavapestov** [comment]
@jrose-apple Override thunks get emitted in each translation unit that has a vtable for a subclass of the class. So you're still relying on the original method having public linkage.

**Comment by huonw** [comment]
@swift-ci please test

**Comment by huonw** [comment]
@swift-ci please test source compatibility

**Comment by huonw** [comment]
@swift-ci please test

@slavapestov I've changed to not setting the subclass scope on thunks, except for signature-optimised thunks, since in that case the original entry point to the code gets marked as a thunk, and thus that needs to have the right subclass scope, so that it remains public if it is `private`.

**Comment by huonw** [comment]
@swift-ci please smoke test

File: include/swift/SIL/SILFunction.h [hunk 2]
```diff
@@ -573,13 +573,27 @@ class SILFunction
 
   /// Get this function's thunk attribute.
   IsThunk_t isThunk() const { return IsThunk_t(Thunk); }
-  void setThunk(IsThunk_t isThunk) { Thunk = isThunk; }
+  void setThunk(IsThunk_t isThunk, bool forSignatureOptimization = false) {
+    // The _original_ function for a method can turn into a thunk through
+    // signature optimization, meaning it needs to retain its subclassScope
+    assert(getClassSubclassScope() == SubclassScope::NotApplicable ||
```
Comment (slavapestov) [inline include/swift/SIL/SILFunction.h]: This seems overkill. If you're intent on keeping this, you should add a new case to IsThunk_t for signature-optimized thunks. A future assertion could also ensure that any thunk other than a signature thunk is shared or private, etc.

**Review by slavapestov — APPROVED** [review]

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please test source compatibility

**Pull request merged**